### PR TITLE
Copy expression package out of /e

### PR DESCRIPTION
**Pull request by EdwardDowling**
Copying expression package out of /e for use in #36745 
There will be a follow up /e PR i'll link here to update /e and remove the package from there

### CODE REVIEW

**Review by codingllama — COMMENTED** [review]
Looks alright to me.

FYI @nklaassen.

File: lib/expression/parser.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils/parse"
+	"github.com/gravitational/teleport/lib/utils/typical"
+)
+
+type evaluationEnvVar map[string]typical.Variable
+
+func DefaultParserSpec[evaluationEnv any]() typical.ParserSpec {
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/expression/parser.go:15]: Is there anything more we could say? What is the "default parser spec"? Is there a definition we could link to, or some other explanation?

**Review by codingllama — APPROVED** [review]

File: lib/expression/parser.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (EdwardDowling) [reply lib/expression/parser.go:16]: > Is there anything more we could say? What is the "default parser spec"? Is there a definition we could link to, or some other explanation?
Added in some more info.

File: lib/expression/parser.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ import (
 type evaluationEnvVar map[string]typical.Variable
 
 // DefaultParserSpec is the default parser specification.
+// Contians useful functions for manipulating and comparing strings.
```
Comment (codingllama) [inline lib/expression/parser.go:15]:
```suggestions
// Contains useful functions for manipulating and comparing strings.
```

**Review by codingllama — APPROVED** [review]

**Review by nklaassen — APPROVED** [review]

**Pull request merged**